File: c#/GrpcBenchmark/Server/Services/GreeterService.cs
using Google.Protobuf;
using Google.Protobuf.WellKnownTypes;
using Grpc.Core;
using Greet;
using System.Diagnostics;

namespace GrpcGreeter.Services;

public class GreeterService : Greeter.GreeterBase
{
    private static Dictionary<uint, ByteString> responseMap;

    static GreeterService()
    {
        responseMap = new Dictionary<uint, ByteString>();
        for (uint size = 10; size <= 1000000; size *= 10)
        {
            responseMap.Add(size, ByteString.CopyFrom(new byte[size]));
        }
    }


    private readonly ILogger<GreeterService> _logger;
    public GreeterService(ILogger<GreeterService> logger)
    {
        _logger = logger;
    }

    public override Task<HelloReply> SayHello(HelloRequest request, ServerCallContext context)
    {
        return Task.FromResult(new HelloReply
        {
            Message = "Hello " + request.Name
        });
    }

    public override Task<DownloadReply> Download(DownloadRequest request, ServerCallContext context)
    {
        if (!responseMap.TryGetValue(request.RequestSize, out ByteString? byteString))
        {
            // 1 KB by default
            byteString = responseMap[1000];
        }

        return Task.FromResult(new DownloadReply
        {
            Body = byteString
        });
    }

    public override async Task DownloadStream(IAsyncStreamReader<DownloadRequest> requestStream, IServerStreamWriter<DownloadReply> responseStream, ServerCallContext context)
    {
        while (await requestStream.MoveNext() && !context.CancellationToken.IsCancellationRequested)
        {
            uint requestSize = requestStream.Current.RequestSize;
            if (!responseMap.TryGetValue(requestSize, out ByteString? byteString))
            {
                // 1 KB by default
                byteString = responseMap[1000];
            }

            await responseStream.WriteAsync(new DownloadReply
            {
                Body = byteString
            });
        }
    }

    public override Task<Empty> Sleep(SleepRequest request, ServerCallContext context)
    {
        // use sleep to mimic blocking operation
        Thread.Sleep((int)(1000 * request.Seconds));
        int threadId = Thread.CurrentThread.ManagedThreadId;
        _logger.LogInformation($"Thread id: {threadId}");
        return Task.FromResult(new Empty());
    }
}

File: c#/GrpcBenchmark/grpc-benchmark-rust/server/src/greeter_service.rs
use proto::{DownloadReply, DownloadRequest, Greeter, HelloReply, HelloRequest, SleepRequest};
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::RwLock;
use tonic::{Request, Response, Status};
use tracing::info;

#[derive(Default)]
pub struct GreeterService {
    response_map: Arc<RwLock<HashMap<u32, Vec<u8>>>>,
}

impl GreeterService {
    async fn get_byte_string(&self, size: u32) -> Vec<u8> {
        let mut map = self.response_map.write().await;
        
        if let Some(bytes) = map.get(&size) {
            bytes.clone()
        } else {
            let bytes = vec![0u8; size as usize];
            map.insert(size, bytes.clone());
            bytes
        }
    }
}

#[tonic::async_trait]
impl Greeter for GreeterService {
    async fn say_hello(
        &self,
        request: Request<HelloRequest>,
    ) -> Result<Response<HelloReply>, Status> {
        let name = request.into_inner().name;
        let reply = HelloReply {
            message: format!("Hello {}", name),
        };
        Ok(Response::new(reply))
    }

    async fn download(
        &self,
        request: Request<DownloadRequest>,
    ) -> Result<Response<DownloadReply>, Status> {
        let request_size = request.into_inner().request_size;
        let body = self.get_byte_string(request_size).await;
        
        let reply = DownloadReply { body };
        Ok(Response::new(reply))
    }

    type DownloadStreamStream = tonic::codec::Streaming<DownloadReply>;

    async fn download_stream(
        &self,
        request: Request<tonic::Streaming<DownloadRequest>>,
    ) -> Result<Response<Self::DownloadStreamStream>, Status> {
        let mut stream = request.into_inner();
        let (tx, rx) = tokio::sync::mpsc::channel(128);

        tokio::spawn(async move {
            while let Some(request) = stream.message().await.unwrap_or(None) {
                let request_size = request.request_size;
                let body = vec![0u8; request_size as usize];
                
                let reply = DownloadReply { body };
                if tx.send(Ok(reply)).await.is_err() {
                    break;
                }
            }
        });

        Ok(Response::new(tonic::codec::Streaming::new(rx)))
    }

    async fn sleep(
        &self,
        request: Request<SleepRequest>,
    ) -> Result<Response<()>, Status> {
        let seconds = request.into_inner().seconds;
        
        // Use sleep to mimic blocking operation
        tokio::time::sleep(tokio::time::Duration::from_secs(seconds as u64)).await;
        
        let thread_id = std::thread::current().id();
        info!("Thread id: {:?}", thread_id);
        
        Ok(Response::new(()))
    }
} 